### ISSUE TITLE
Need more complete list of example questions

### ISSUE THREAD

**Issue by edeutsch** [RTXteam/RTX#118]
We need to have a list of example questions that can be asked. We need this for two purposes. First, the example list of questions that could be asked is now woefully small at the main UI: http:/rtx.ncats.io/ and we need to augment that. Second, we need to give Arnab et al a list of questions for their autocomplete.

We already have a document for query types:
https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/QuestionAnswering/Questions.tsv

What's the best way to generate a list of questions?
a) We could just add another column manually that is one instance of a question for that query type
b) We could create a new file of questions, perhaps two columns: query type id, question. The advantage with a separate file is that we could have several questions per query type (more easily) and it might be easier to edit that Questions.tsv, which is getting quite wide (and is mis-named according to controlled vocabulary)
c) We could try to autogenerate such a list based on the query types and nodes in the graph. This is feasible and maybe useful for the autocomplete. However, we run the risk of suggesting questions for which there is not yet a reasonable answer, which looks bad. So maybe better to have the list we hand-curated at the start at least.

Comments? Thoughts? Already started somewhere? Someone willing to tackle this? I can come up with the design, but I don't feel qualified to select nice examples of each query type.

@dkoslicki @isbluis Jared?

**Comment by dkoslicki** [comment]
There is some code to do this currently:
[WhatQuestionsCanYouAnswer.py](https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/QuestionAnswering/WhatQuestionsCanYouAnswer.py) parses ``Questions.py``, pulls random nodes from the KG, and then populates the query templates with the correct node types.

Of course, since it's pulling random nodes, there are drawbacks (like the ones you mentioned with questions not having a reasonable answer).

**Comment by dkoslicki** [comment]
I'm definitely open to a manually curated list of example questions. We just need someone to gather them. Perhaps this is a reasonable task for @jaredroach (since we need an SME to assess if the answers are "reasonable")?

**Comment by edeutsch** [comment]
Great. okay, @jaredroach, would you complete this file for example questions:
https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/QuestionAnswering/QuestionExamples.tsv
You can edit directly with the little pencil icon.

Key things:
- the first and second column must be separated by a TAB
- The first column Qnn refers to a specific known_query_type_id. Do not re-number.
- It is okay for there to be more than one question for a given known_query_type_id. One question per line. See examples.
- All cases of "$entity_type" should be replace by a specific instance of that type. For example, $disease could become malaria
- Each known_query_type_id should have at least one example
- Ideally these examples would be reasonable questions to ask with a fairly reasonable response. Doesn't have be to an awesome answer, but something non-embarassing or non-existent.

Please respond to this thread if you have questions.

thanks!

**Comment by jaredroach** [comment]
This seems like it might be really frustrating for me.
For example, this seems like about as "softball" a question I can ask the Reasoner for Q4, and yet...

http://rtx.ncats.io
Enter question: 
what proteins are in the glycolysis pathway
Status
The most similar question I can answer is: what proteins are in the pathway $pathway But I was unable to fill the following parameters: ['pathway']

Swapping the order of the words 'glycolysis' and 'pathway' got the same result.

and then matching the template with quotes got this result:
what proteins are in the pathway 'glycolysis'
Interpreting your question...
An error was encountered:
INTERNAL SERVER ERROR (500)

**Comment by edeutsch** [comment]
Well, I guess this is a ParseQuestion bug for @dkoslicki . I hope very few of them will be frustrating in this way, but I have not tested these. I've noticed that David has been generating lots of Qnns but I've just been testing with what you see on the main interface. The main goal of this is to augment what we see on the web page.

@dkoslicki , should Jared wait on this until you do the brain transplant over the next couple days?

**Comment by dkoslicki** [comment]
@jaredroach There does not appear to be a ``glycolysis`` pathway in the knowledge graph:
on http://rtx.ncats.io:7474/browser/
```
match (n:pathway) where n.description="glycolysis" return n
```
return no matches.

**Comment by dkoslicki** [comment]
Also, avoid the use of single quotes/apostrophes in the UI pending a fix to #121

**Comment by jaredroach** [comment]
Is there something I need to do to connect to neo4j?

http://rtx.ncats.io:7474/browser/
match (n:pathway) where n.description="glyc" return n
BoltConnectionError
No connection found, did you connect to Neo4j?

**Comment by jaredroach** [comment]
We do have Reactome pathways. And a node for
http://reactome.org/content/detail/R-HSA-70326
which is named "Glucose metabolism" in Reactome.
Reactome lists its synonyms as "Glycolysis and gluconeogenesis"
So we would be well served to be able to recognize Reactome synonyms if we are expecting this type of question.
Or maybe include more of REactome? Because it appears that Reactome R-HAS-70171 is 'glycolysis' but is not in the KG?

So I was able to ask this question on Safari
What proteins are part of the pathway Glucose metabolism?
and it seemed like the interface was working, and after about 5 minutes, results started to appear. But then something with the interface gobbled up almost all my CPU and Safari started to crawl. So I had to force quit Safari.
I then tried Firefox. Similar thing happened. Firefox actually asked me "A web page is slowing down your browser. What would you like to do?"
I let it run and after about a minute or so, 1560 results were found.

So I guess that 1560 results was the problem. It would be nice if the answer wasn't so promiscuous. In this case, it is too promiscuous for two reasons:
1. So many results that the browser is so slow it is almost impossible for the user to glacially interact with the results.
2. So many results that the user is overwhelmed. A human usually doesn't want more than a few dozen results maximum.

There should be <<100 "true" results: e.g., the enzymes that appear in Figure 3.1-1. of Michal's Biochemical Pathways "Glycolysis and Gluconeogenesis".

**Comment by dkoslicki** [comment]
@edeutsch @isbluis Do you want to handle the trimming of results to only the top 100, or should I do this in the reasoning code itself? I can imagine that we want the API to return the full results (all 1560 in this case), but the UI may want to trim this for performance purposes as suggested by @jaredroach

**Comment by jaredroach** [comment]
Are we sure there are 1560 true results? Surely some of those are false positives? That even the API wouldn't want to return...

**Comment by jaredroach** [comment]
I tried searching for a simpler pathway that might produce fewer protein members.
I made sure it was in Reactome before searching
What proteins are in  the pathway Thyroxine biosynthesis?
but apparently http://reactome.org/content/detail/R-HSA-209968 is not in the KG either.

**Comment by isbluis** [comment]
From the UI perspective, it is not a great idea to hand it a HUGE json object.  Just return the top N (100? user-settable?) and a way to page, if desired.

**Comment by edeutsch** [comment]
I could implement a max_results=nn parameter to /query that the GUI to could set to some reasonable default? Does that sound like a good solution? If we wanted to go full paging now, that could be done as well. page_size=nn, page_number=nn. Although that would require more on the GUI and maybe not worth it at this stage?

**Comment by dkoslicki** [comment]
@edeutsch I think a max_result=nn is fine for the UI.

@jaredroach Yes, the KG really does say there are 1,560 proteins in that pathway. I cannot assess the accuracy since this is coming directly from the KS that @saramsey is parsing for protein->pathway relationships (and they don't have confidence scores associated to them).

**Comment by jaredroach** [comment]
So we may have a problem answering any questions related to pathway content if all the pathways have as many false positives as glycolysis. Maybe a task to prune the KG?

In the meantime, my idea is to try to find at least one pathway that has a reasonably small number of proteins. Is there a cypher command that can return all the pathway nodes that have exactly 5 members? Something like the following (which doesn't work):

start n=node(*)
 match n-[r]-()
 return n, count(r) where count(r) = 5

My idea would then be to scan the list to see if any of these pathways are known well enough by the average user to be good example questions.

**Comment by jaredroach** [comment]
Thanks. I set n= 10 and got 4 results. One of the pathways was recognizable to an average user, "mRNA Editing". So I tried "What proteins are in the mRNA editing pathway?" and got the ten results back. So far, so good. But it turns out these 10 proteins are not really in a "pathway". Rather this a thematic set of proteins that are involved in mRNA editing, but each in a potentially distinct pathway. So I rephrased the question to "What are proteins are involved in mRNA editing?" and the restated question was identical to the first query I posed, and the answer was the same. So the system is getting a correct answer to "What proteins are involved in mRNA editing?" but an 'incorrect' answer to "What proteins are in the mRNA editing pathway?" because the question is poorly phrased (if indeed any answer to such a question can be either correct or incorrect).

So - question to @edeutsch - for https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/QuestionAnswering/QuestionExamples.tsv - does "What proteins are involved in mRNA editing?"  serve as an example for the template "what proteins are in the pathway $pathway" or not? Or do we need to find an example of a real pathway? fwiw, most of the nodes labeled as "pathways" in the KG are not really pathways, IMHO. They are more like Gene Ontology terms.

**Comment by edeutsch** [comment]
Well, I suppose I would pass this question to @dkoslicki and @saramsey for comment. It's possible that the query strategy is not correct. Or maybe the query strategy is correct but the KG really isn't correct, perhaps because the knowledge source has poor data. I think it would be worth deconstructing exactly how we get this answer and see if we have a bug somewhere or misinterpretation or if we're doing everything right but the KS is poor.

It might be worth filing this topic as a separate bug report, since it would seem that we have a serious deficiency in our pathway data that should be studied and remedied. Just for fun I took a peek to see what we did in PeptideAtlas:
https://db.systemsbiology.net/sbeams/cgi/PeptideAtlas/showPathways
(set build to the 2018-01b if it isn't already)
The top pathway is Glycolysis / Gluconeogenesis. If you click on the PA symbol next to it, you see the KEGG pathway with proteins we have in PeptideAtlas in green. Mouse over for the GeneIDs. How does this square with what answer you would expect to receive from that kind of question.

So, in ultimate answer to your question, I would open up a separate ticket to explore this further, because this seems like an important type of question to get right. I think at the moment we don't have a good candidate for a "good example" of this question and we should flag the problem and continue on to others. I would hope that most of these should be pretty easy and you just stumbled into a bad one at the beginning. We'll see. 

So, I would leave the entry as is with a "$pathway" and move on to the next question. Maybe you can ignore all the $pathway questions until we explore this further.

**Comment by edeutsch** [comment]
..and I just tried caffeine metabolism and necroptosis as two more likely candidates but with no useful response. And I just notice that there's also a fanconi anemia pathway:
https://db.systemsbiology.net/sbeams/cgi/PeptideAtlas/showPathways?apply_action=pathway_details;path_id=path:hsa03460;path_def=Fanconi%20anemia%20pathway%20-%20Homo%20sapiens%20%28human%29;searchdb=pa

**Comment by dkoslicki** [comment]
@jaredroach The question answering is based off of template matching: it finds the closest match in ``Questions.tsv`` to the input question and tries to populate the relevant term(s). "What are proteins are involved in mRNA editing?" and "What proteins are in the mRNA editing pathway?" both match to Q4. If you ask a question that is not in ``Questions.tsv`` it will either:
1. Throw an error
2. Match a related question (if it [semantically close enough](https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/QuestionAnswering/WordnetDistance.py))

As for the proteins in the pathways you mention, these come from Reactome. If you are observing errors, this is either:
1. An issue with reactome
2. An issue with the KG construction (and @saramsey should be consulted).

From a reasoning perspective, if something is called a pathway in the KG, I treat it as a pathway. If a protein participates in a pathway, I take this as truth (as there is no other way I can ascertain the veracity of this claim). Steve can comment about the accuracy of the data.

**Comment by dkoslicki** [comment]
So as for @edeutsch's comment: I can confirm that the query strategy is correct (it's a simple ``match (s:protein)-[r:participates_in]-(t:pathway) return t`` query).

**Comment by edeutsch** [comment]
Maybe Reactome is the problem? Maybe we should be using KEGG pathways instead? or in addition? Our use of KEGG in PeptideAtlas seems to return what seems like quite nice results.

**Comment by jaredroach** [comment]
I don't see how reactome is the problem. Maybe we are following too many links from the Reactome raw database files. If I go to this page for glycolysis:
https://reactome.org/content/detail/R-HSA-70171
I don't see how one gets more than a few dozen proteins.
KEGG (eg the graph @edeutsch linked to) also shows at most a few dozen proteins. 

Somehow the "thousands" number of proteins involved in the pathway must somehow include any protein vaguely influenced by glycolysis. Which is going to be quite a lot because glycolysis provides energy to build these proteins and drive their reactions. We don't want these in the Reasoner. Unless they have special edges in the KG which are ignored for most reasoning queries. These edges might be used for a question like "Really go out on a limb and don't give me the obvious answers, but rather tell me what proteins might be conceivably influenced by glycolysis?" Importantly, any SME at the Hackthon is going to consider these promiscuous answers as incorrect. So we don't want these KG edges for now - not for the minimally viable product (MVP).

**Comment by jaredroach** [comment]
The phenotypes question template is also going to have problems:

 What are the phenotypes of sickle cell anemia?
Status
Your question has been interpreted and is restated as follows:
   what are the phenotypes of the disease SICKLE CELL ANEMIA?
Please ensure that this is an accurate restatement of the intended question.

No path between OMIM:603903 and phenotypic_feature via relationship has_phenotype


BUT THE GOOD NEWS is that
What are the phenotypes of Fanconi anemia?
works pretty well

**Comment by jaredroach** [comment]
What genes are expressed in tendons?
cannot be interpreted.
But What genes are expressed in the tendon?
is properly interpreted.

But the first phrasing is more standard.

Note that pretty much all the what genes are expressed in $anatomical_entity
crash my browser as they return too many results. Retina was 580 results.

**Comment by jaredroach** [comment]
We seem to have problems with this template as well

what proteins are associated with $phenotypic_feature
What proteins are associated with cryptorchidism?
No path between HP:0000028 and protein via relationship associated_with_condition

What proteins are associated with jaundice?
No path between HP:0000952 and protein via relationship associated_with_condition

I went with 
What proteins are associated with anemia?
Your question has been interpreted and is restated as follows:
   what genes are associated with anemia?
Please ensure that this is an accurate restatement of the intended question.

26 result found
but it rephrased it with genes, not proteins (and it may not be thinking of anemia as a phenotype, but rather a condition)


--

**Comment by jaredroach** [comment]
We fail with the 
what micrornas are expressed in $anatomical_entity
template
e.g., What micrornas are expressed in the blood? What micrornas are expressed in the retina? 
because it reinterprets the question as
What genes are expressed in $anatomical_entity

**Comment by jaredroach** [comment]
for the template

what diseases involve similar genes to $disease
What diseases involve similar genes to Fanconi anemia?
gives one result
The disease leukopenia involves similar protein's as Fanconi anemia with similarity value 0.272727
which is a pretty cool and biological interesting result
EXCEPT
that leukopenia is a phenotype, NOT a disease

the question
What diseases involve similar genes to sickle cell anemia?
gives a bunch of interesting results to this question
EXCEPT the top hit TO ITSELF, with a score <<1
The disease sickle cell anemia involves similar protein's as SICKLE CELL ANEMIA with similarity value 0.346154
and (minor point) related to the leukopenia point above. One of the retuned hits is "anemia" which is a phenotype not a disease
-

**Comment by jaredroach** [comment]
What drugs have similar protein targets to naproxen?
gives a pretty nice answer. If you had to pick a small set of questions to showcase, this might be one of them.

**Comment by dkoslicki** [comment]
Per [this question](https://github.com/RTXteam/RTX/issues/118#issuecomment-388194301): some diseases have no phenotypes (as is the case with this one). @saramsey will be the one to comment on if this is a KG bug or limitation of the KS's we're using.

[this comment](https://github.com/RTXteam/RTX/issues/118#issuecomment-388197059) Opened an issue #134

[this comment](https://github.com/RTXteam/RTX/issues/118#issuecomment-388198509) same statement about KG completeness (i.e. may be a bug in KG construction, or limitation of KS used). Our KG treats genes and proteins as identical (as per @saramsey's design choice).

[this comment](https://github.com/RTXteam/RTX/issues/118#issuecomment-388199030) this is a known issue on my end. Problem is that "microrna" is not an English word according to ``nltk`` so it ignores it. Will investigate a fix eventually.

[this comment](https://github.com/RTXteam/RTX/issues/118#issuecomment-388200775) According to Disease Ontology, leukopenia is a disease, so this is an issue with the KS. Maybe it will change when we switch to MONDO. Curiously, when I run that question on FA, I get:
```
DOID:10485      esophageal atresia      Jaccard 0.290323
DOID:615        leukopenia      Jaccard 0.272727
```

**Comment by jaredroach** [comment]
I am having problems with coming up with an example for this template:
what proteins are members of similar pathways to $protein

Either the question hangs (presumably at the server end, as no results ever get to my browser) OR the protein cannot be found OR the protein has no links to pathways.

What proteins are members of similar pathways as PTGS1?
-> hangs

What proteins are members of similar pathways as DHODH?
-> hangs

What proteins are members of similar pathways as BCL11A?
Your question has been interpreted and is restated as follows:
   what proteins are members of similar pathways to BCL11A?
Please ensure that this is an accurate restatement of the intended question.
Sorry, the protein Q9H165 is not connected to any pathway.

What proteins are members of similar pathways as LRF?
The most similar question I can answer is: what proteins are members of similar pathways to $protein But I was unable to fill the following parameters: ['protein']
-> cannot find a synonym for LRF

What proteins are members of similar pathways as ZBTB7A?  (aka LRF)
The most similar question I can answer is: what proteins are members of similar pathways to $protein But I was unable to fill the following parameters: ['protein']
-> see next example

What proteins are members of similar pathways as O95365?
Your question has been interpreted and is restated as follows:
   what proteins are members of similar pathways to ZBT7A?
Please ensure that this is an accurate restatement of the intended question.
Sorry, the protein O95365 is not connected to any pathway.
-> cannot find a synonym for ZBTB7A as O95365, but CAN find the synonym in the opposite direction ( O95365 to ZBTB7A) & then even with this success, cannot connect it to any pathway.

**Comment by jaredroach** [comment]
I am having problems with coming up with an example for this template:
what proteins are expressed in similar anatomy to $protein

either I get "undefined" for an answer, or it cannot find a synonym (even for hemoglobin)

#hair follicle
#MKAP13
What proteins are expressed in similar anatomy to Q8IUC0?
What proteins are expressed in similar anatomy to Q8IUC0?
Your question has been interpreted and is restated as follows:
   what proteins are expressed in similar anatomy to KRTAP13-1?
Please ensure that this is an accurate restatement of the intended question.
Looking for answer...
An error was encountered:
undefined

What proteins are expressed in similar anatomy to RPE65?
Your question has been interpreted and is restated as follows:
   what proteins are expressed in similar anatomy to RPE65?
Please ensure that this is an accurate restatement of the intended question.
Looking for answer...
An error was encountered:
undefined

What proteins are expressed in similar anatomy to hemoglobin?
Status
The most similar question I can answer is: what proteins are expressed in similar anatomy to $protein But I was unable to fill the following parameters: ['protein']

What proteins are expressed in similar anatomy to PTGS1?
Your question has been interpreted and is restated as follows:
   what proteins are expressed in similar anatomy to PTGS1?
Please ensure that this is an accurate restatement of the intended question.
Looking for answer...
An error was encountered:
undefined

**Comment by jaredroach** [comment]
For this template
what proteins are associated with similar phenotypes to $protein

#retinitis pigmentosa
What proteins are associated with similar phenotypes to USH2A?
gives a great response

HOWEVER, this one for Fanconi anemia fails
What proteins are associated with similar phenotypes to FANCD2?
Your question has been interpreted and is restated as follows:
   what proteins are associated with similar phenotypes to FANCD2?
Please ensure that this is an accurate restatement of the intended question.
Sorry, the protein Q9BXW9 is not connected to any phenotypic_feature.
--

**Comment by jaredroach** [comment]
Capitalization
Would be nice if the reworded question was capitalized. It is jarring to see un-capitalized sentences.

What proteins are expressed in similar anatomy to PTGS1?
Your question has been interpreted and is restated as follows:
what proteins are expressed in similar anatomy to PTGS1?

**Comment by jaredroach** [comment]
This question
What genetic conditions involve similar micrornas to retinitis pigmentosa?
leads to:

Interpreting your question...

An error was encountered:
INTERNAL SERVER ERROR (500)

and stayed hung like that on one trial. I tried again in a new browser window and got the same message for about one second and then it was updated to:
Your question has been interpreted and is restated as follows:
   what genetic conditions involve similar micrornas to DOID:10584?
Please ensure that this is an accurate restatement of the intended question.
Sorry, the disease DOID:10584 is not connected to any microRNA.

**Comment by dkoslicki** [comment]
[For this question](https://github.com/RTXteam/RTX/issues/118#issuecomment-388454392), we have duplicate nodes from OMIM and DOID. @saramsey Do we want to address the fact that both DOID:10584 and OMIM:268000 are in the graph (and have different associated edges)?

**Comment by dkoslicki** [comment]
@jaredroach [As for the hanging issue](https://github.com/RTXteam/RTX/issues/118#issuecomment-388451250), I have updated the code and it is running significantly faster. See 4c29c31

**Comment by jaredroach** [comment]
This question works and can be interpreted:
What pathways involve similar genes to the phenylketonuria pathway?

However these cannot (which may be a recent break - I thought it worked last week):
What proteins are in the phenylketonuria pathway?
What drugs target proteins in the phenylketonuria pathway?
-> Status
The most similar question I can answer is: What proteins are in the pathway $pathway But I was unable to fill the following parameters: ['pathway']
--

**Comment by jaredroach** [comment]
Trying to find a question to match this template:
What drugs target proteins associated with $anatomical_entity

but surprisingly this doesn't work:
What drugs target proteins associated with the kidney?
Status
The most similar question I can answer is: What drugs target proteins associated with $disease But I was unable to fill the following parameters: ['disease']

--

**Comment by dkoslicki** [comment]
re [this question](https://github.com/RTXteam/RTX/issues/118#issuecomment-389333706), the question has been restated as "what drugs target proteins expressed in $anatomical_entity" to more closely mimic the structure in the KG (proteins in the KG are related to anatomic_entities with "is_expressed_in" not "is_associated_with"). @jaredroach If you want, I could also make it work with "associated" if you feel that's desired

**Comment by dkoslicki** [comment]
re [this question](https://github.com/RTXteam/RTX/issues/118#issuecomment-389333412) opening a separate issue for this error.

**Comment by jaredroach** [comment]
I am having a hard time finding a question of the form
What drugs target proteins associated with $biological_process
that works.

What drugs target proteins associated with synaptic function?
What drugs target proteins associated with meiosis?
What drugs target proteins associated with mitosis?

all fail as the interpreter tries to find a disease rather than a biological process.
The most similar question I can answer is: What drugs target proteins associated with $disease But I was unable to fill the following parameters: ['disease']

**Comment by jaredroach** [comment]
It is not clear what the distinction is between biological process and pathway. Consider this response:
"The biological_process cyclooxygenase pathway involves similar protein's as naproxen with similarity value 0.096774"

**Comment by jaredroach** [comment]
What proteins involve similar biological processes as tubulin?
->Your question has been interpreted and is restated as follows:
   What proteins involve similar biological processes as Tubulin?
Please ensure that this is an accurate restatement of the intended question.
Sorry, the protein A6NNZ2 is not connected to any biological_process.

Kind of surprising. Tubulin is so central.

**Comment by dkoslicki** [comment]
From @jaredroach in a different comment thread:

> neither of these questions can be interpreted:
What drugs target proteins associated with scars?
What drugs target proteins associated with keloid?
>
> maybe "scar" or "keloid" should be in the KG
